cmds/switch_root/switch_root.go

```diff
@@ -23,140 +26,171 @@ func usage() string {
 	return "switch_root [-h] [-V]\nswitch_root newroot init"
 }
 
-// littleDoctor recursively deletes everything at "path" that
-// is in the same file system as "fs".
-func littleDoctor(path string, fsType int64) error {
-	var pathFS syscall.Statfs_t
+// get device for file descriptor
+func getDev(fd int) (dev uint64, err error) {
+	var stat syscall.Stat_t
 
-	if err := syscall.Statfs(path, &pathFS); err != nil {
-		return err
+	if err := syscall.Fstat(fd, &stat); err != nil {
+		return 0, err
 	}
 
-	if int64(pathFS.Type) != fsType {
-		return nil
+	return stat.Dev, nil
+}
+
+// recursively deletes a directory and everything in it
+// works with a file descriptor, can delete files not referenceable
+// any more (e.g. after chroot)
+// does not descent into mounts
+func recursiveDelete(fd int) error {
+	parentDev, err := getDev(fd)
+	if err != nil {
+		return err
 	}
 
-	file, err := os.Open(path)
+  // the file descriptor is already open, but allocating a os.File
+	// here for it makes reading the files in the dir so much nicer
+	dir := os.NewFile(uintptr(fd), "__ignored__") // filename is completely irrelevant
+	defer dir.Close()
+	names, err := dir.Readdirnames(-1)
 	if err != nil {
-		return fmt.Errorf("Could not open %s: %v", path, err)
+		return err
 	}
 
-	if fileStat, err := file.Stat(); fileStat.IsDir() {
-		if err != nil {
+	for _, name := range names {
+		// loop here, but handle loop in separate function to make defer work as expected
+		if err := recusiveDeleteInner(fd, parentDev, name); err != nil {
 			return err
 		}
+	}
+	return nil
+}
 
-		names, err := file.Readdirnames(-1)
-		defer file.Close()
-		if err != nil {
+// don't call this directly, this function is the loop content of recursiveDelete
+func recusiveDeleteInner(parentFd int, parentDev uint64, childName string) error {
+	// O_DIRECTORY and O_NOFOLLOW make this open fail for all files and all symlinks (even when pointing to a dir)
+	// we need to filter out symlinks because getDev later follows them
+	childFd, err := syscall.Openat(parentFd, childName, syscall.O_DIRECTORY | syscall.O_NOFOLLOW, syscall.O_RDWR)
```

Comment:
super nit: let's just use unix everywhere instead of syscall. It has everything but the errnos.